pkg/rpcmiddleware/discover/discover.go

```diff
@@ -77,64 +99,106 @@ func DefaultURL(target string, port int, tls bool) (string, error) {
+// Use this option with gRPC's WithInsecure dial option.
+func WithInsecureFallback() Option {
+	return optionFunc(func(opts *options) {
+		opts.insecureFallback = true
+	})
 }
 
-// DialOptions discovers gRPC dial options based on the given target. This includes whether or not transport level
-// security is enabled and service port discovery.
-func DialOptions(ctx context.Context, target string, creds credentials.TransportCredentials) ([]grpc.DialOption, error) {
-	// TODO: Discover through SRV records and cache result (https://github.com/TheThingsNetwork/lorawan-stack/issues/138)
-	if val, ok := ctx.Value(tlsFallbackKey).(bool); ok && !val {
-		return WithInsecure(), nil
-	}
-	return WithTransportCredentials(creds), nil
+// WithDNSResolver configures the dialer with a custom DNS resolver.
+func WithDNSResolver(resolver DNSResolver) Option {
+	return optionFunc(func(opts *options) {
+		opts.dnsResolver = resolver
+	})
 }
 
-// DialContext creates a client connection to the given target. It uses DialOptions to discover gRPC dial options for
-// the target.
-func DialContext(ctx context.Context, target string, creds credentials.TransportCredentials, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	discoveredOpts, err := DialOptions(ctx, target, creds)
-	if err != nil {
-		return nil, err
+// WithAddressDialer configures the discovery dialer with a custom address dialer.
+// This dialer is called for each discovered address until a nil error is returned.
+func WithAddressDialer(dialer func(context.Context, string) (net.Conn, error)) Option {
+	return optionFunc(func(opts *options) {
+		opts.dialer = dialer
+	})
+}
+
+// WithDialer returns a gRPC dialer that uses service discovery.
+//
+// Service discovery performs a DNS SRV lookup on the given target if it does not contain a port and the given role is
+// discoverable.
+//
+// If the given role does not support discovery or if the DNS SRV lookup fails, this function falls back to the default port.
+func WithDialer(role ttnpb.ClusterRole, opts ...Option) grpc.DialOption {
+	options := &options{
+		insecureFallback: false,
+		dnsResolver:      net.DefaultResolver,
+		dialer: func(ctx context.Context, address string) (net.Conn, error) {
+			return new(net.Dialer).DialContext(ctx, "tcp", address)
+		},
+	}
+	for _, opt := range opts {
+		opt.apply(options)
 	}
-	allOpts := make([]grpc.DialOption, 0, len(discoveredOpts)+len(opts))
-	allOpts = append(allOpts, discoveredOpts...)
-	allOpts = append(allOpts, opts...)
-	return grpc.DialContext(ctx, target, allOpts...)
+
+	return grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
+		var addresses []string
+		if _, _, err := net.SplitHostPort(target); err == nil {
+			addresses = []string{target}
+		} else {
+			if service, ok := ServiceName(role); ok {
+				_, addrs, err := options.dnsResolver.LookupSRV(ctx, service, "tcp", target)
+				if err == nil {
```

Comment:
Shouldn't we return lookup errors (other than not found)?

```go
var dnsErr *net.DNSError
if errors.As(err, &dnsErr) {
	if !dnsErr.IsNotFound {
		return nil, err
	}
}
```

Reply:
84617e29b